[MLIR] Fix verifier for TF::Conv2DOp and TF::Conv3DOp.

pr4tgpt opened the pull request:
This commit addresses a bug where in tf.conv2D and tf.conv3D, incorrect sizes of output shape was not verified. Added changes to verify the output shape and emit error for incorrect output shapes.

Signed-off-by: Prateek Gupta <prateek@polymagelabs.com>

pr4tgpt committed:
[MLIR] Fix verifier for TF::Conv2DOp and TF::Conv3DOp.
This commit addresses a bug where in tf.conv2D and tf.conv3D, incorrect sizes of output shape was not verified. Added changes to verify the output shape and emit error for incorrect output shapes.

Signed-off-by: Prateek Gupta <prateek@polymagelabs.com>

bondhugula commented:
@smit-hinsu for visibility - as the author of the surrounding code.

smit-hinsu commented:
Thanks for the improvements!

Could you make use the new inferReturnTypes trait for shape inference?

See this commit for an example. https://github.com/tensorflow/tensorflow/commit/edeb46930e7e6cef0154ef7fa8014bb42d4052c2

pr4tgpt commented:
> Thanks for the improvements!
> 
> Could you make use the new inferReturnTypes trait for shape inference?
> 
> See this commit for an example. [edeb469](https://github.com/tensorflow/tensorflow/commit/edeb46930e7e6cef0154ef7fa8014bb42d4052c2)

It isn't immediately clear from the other revision as to how inferReturnTypes should be used/incorporated. Could you please expand a bit more on what changes/additions should be made here? Thanks

smit-hinsu commented:
Once we declare InferTypeOpInterface interface and implement inferReturnTypes methods for these ops, we don't need to explicitly call shape verification. You should see that shapes are being verified automatically without explicit call to inferReturnTypes. The benefit of this generic approach is that the interface can be shared between buillder, verifier and allows other uses as all the ops follow the same mechanism.

Checkout interfaces documentation at https://mlir.llvm.org/docs/Interfaces/.

Let me know if you need further explanation or specific details.

pr4tgpt commented:
> Once we declare InferTypeOpInterface interface and implement inferReturnTypes methods for these ops, we don't need to explicitly call shape verification. You should see that shapes are being verified automatically without explicit call to inferReturnTypes. The benefit of this generic approach is that the interface can be shared between buillder, verifier and allows other uses as all the ops follow the same mechanism.
> 
> Checkout interfaces documentation at https://mlir.llvm.org/docs/Interfaces/.
> 
> Let me know if you need further explanation or specific details.

@smit-hinsu , there is an issue in using the ``InferTypeOpInterface``. Previously,  when I was using the verifier to check the output type all other parameters like non-negative strides, a correct number of input dimensions/strides etc were checked before only(as I kept my check at the end of the verifier).
But now if I use the ``inferReturnTypes`` method,  the output type is inferred before doing other sanity checks resulting in a segmentation fault. The output type in the case of ``tf.conv2D`` and ``tf.conv3D`` can only be inferred if the attributes are correct.
When I am running my code using ``inferReturnTypes`` with correct attributes, it is running fine. But if there are incorrect attributes they have to be checked first
If I emit errors in those instances, then there will be no use of a verifier. So could you suggest something here?

smit-hinsu commented:
We need to emit error in that case and fail the type inference. See the following op inference in HLO dialect for an example, 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc#L1687

pr4tgpt commented:
> @pr4tgpt Can you please check @smit-hinsu's comments and keep us posted ? Thanks!

Yes I am working on smit's suggestion. Thanks.

pr4tgpt commented:
> We need to emit error in that case and fail the type inference. See the following op inference in HLO dialect for an example,
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/IR/hlo_ops.cc#L1687

Hi @smit-hinsu I have made changes as requested by you. Please review them. Thanks.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1246,6 +1241,139 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
   return success();
 }
 
+LogicalResult Conv2DOp::inferReturnTypes(
```
can we templatize Conv2D and Conv3D similar to how VerifyConvOpAttributes supports both the ops with a common implementation?

pr4tgpt replied:
Done. Thanks for the suggestion.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1246,6 +1241,139 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+  const Value filter = operands[1];
+  const TensorType input_ty = input.getType().cast<TensorType>();
+  const TensorType filter_ty = filter.getType().cast<TensorType>();
+  const StringRef paddings =
```
nit: rename to paddding. padings gives an impression that the variable contains padidng values.

pr4tgpt replied:
Done.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1246,6 +1241,139 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+  const TensorType input_ty = input.getType().cast<TensorType>();
+  const TensorType filter_ty = filter.getType().cast<TensorType>();
+  const StringRef paddings =
+      attributes.get("padding").cast<::mlir::StringAttr>().getValue();
```
we need to check here that attributes contains this attribute and it is of type String before cast.

pr4tgpt replied:
Used OpAdaptor now. Thanks

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1246,6 +1241,139 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+  }
+
+  // The remaining dimensions can be obtained directly.
+  if (format == tensorflow::TensorFormat::FORMAT_NHWC ||
```
can we use utilities from tensorflow/core/util/tensor_format.h here?

pr4tgpt replied:
Done. Thanks

smit-hinsu reviewed (commented):
Sorry for the delay in the response. I somehow missed the update mail. Please ping on the review if you don't get an update within 24 hours.

This looks quite good but I have a few suggestions.

pr4tgpt commented:
Thank you @smit-hinsu for suggestions. Will update the code with relevant changes.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1246,6 +1241,139 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+    mlir::ValueRange operands, mlir::DictionaryAttr attributes,
+    mlir::RegionRange regions,
+    llvm::SmallVectorImpl<mlir::Type>& inferredReturnTypes) {
+  const int64_t num_spatial_dims = 2;
```
Use the auto-generated op adaptor verify method to avoid manual implementation of some of the checks.

For example,
https://github.com/tensorflow/tensorflow/blob/61dcdcb2fc0869c9ee3486aadbe0c8cbb723cf82/tensorflow/compiler/mlir/lite/ir/tfl_ops.cc#L1453

pr4tgpt replied:
Done. Thanks.

gbaned commented:
@pr4tgpt Can you please check @smit-hinsu's comments and keep us posted ? Thanks!

pr4tgpt commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1156,53 +1158,46 @@ static LogicalResult VerifyConvOpAttributes(OpT op, int num_dims) {
+  };
+
+  int64_t strides_size = strides.size();
+  llvm::errs() << strides_size << " " << num_dims << "\n";
```
Done.

pr4tgpt commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1156,53 +1158,46 @@ static LogicalResult VerifyConvOpAttributes(OpT op, int num_dims) {
+  llvm::errs() << strides_size << " " << num_dims << "\n";
+  if (strides_size != num_dims)
+    return (num_dims == 4)
+               ? emitOptionalError(location,
```
Done. Thanks.

pr4tgpt commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1156,53 +1158,46 @@ static LogicalResult VerifyConvOpAttributes(OpT op, int num_dims) {
+static LogicalResult VerifyConvOpAttributesV2(
+    int num_dims, ArrayRef<Attribute> strides, ArrayRef<Attribute> dilations,
+    llvm::Optional<mlir::Location> location) {
+  auto is_not_positive = [](Attribute val) {
```
Done. Thanks

pr4tgpt commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1156,53 +1158,46 @@ static LogicalResult VerifyConvOpAttributes(OpT op, int num_dims) {
   return success();
 }
 
+static LogicalResult VerifyConvOpAttributesV2(
```
Please review the change. Thanks

pr4tgpt commented:
> @pr4tgpt Can you please check @smit-hinsu's comments and keep us posted ? Thanks!

I have checked and resolved the comments. Thanks.

pr4tgpt commented:
Hi, @smit-hinsu , I have made changes as requested by you. Please review the changes and feel free to post suggestions/improvements. Thanks!

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_generated_ops.td:
```diff
@@ -2021,7 +2025,8 @@ Our Conv3D implements a form of cross-correlation.
     DefaultValuedAttr<TF_AnyStrAttrOf<["NDHWC", "NCDHW"]>, "NDHWC">:$data_format,
-    DefaultValuedAttr<I64ArrayAttr, "{1, 1, 1, 1, 1}">:$dilations
+    DefaultValuedAttr<I64ArrayAttr, "{1, 1, 1, 1, 1}">:$dilations,
+    DefaultValuedAttr<I64ArrayAttr, "{}">:$explicit_paddings
```
This information is based on the definition here so we don't have explicit paddings attribute available.
https://github.com/tensorflow/tensorflow/blob/35b14d52e704d316c603dd2599c3c88f329130fe/tensorflow/core/ops/nn_ops.cc#L639

So, we will need to conditionally handle this attribute by fetching attribute based on the name.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1129,80 +1131,50 @@ LogicalResult ConstOp::inferReturnTypes(
 
+template <typename OpT>
+static LogicalResult VerifyConvOpAttributes(OpT op, int num_dims) {
+  if (!IsOfRankOrUnranked(op.getResult(), num_dims))
```
We shouldn't need to verify the result type explicitly as it is done within your method. So, we can rename VerifyConvOpAttributesV2 as VerifyConvOpAttributes and only have one function to verify attributes.

pr4tgpt replied:
Hi @smit-hinsu I have made two versions as ``conv2DBackpropInputOp`` is also using ``verifyConvOpAttributes``, and there the result type is not checked.

smit-hinsu replied:
This seems it hasn't been changed and we still have VerifyConvOpAttributesV2 andd VerifyConvOpAttributes.

smit-hinsu replied:
As you mentioned earlier, we will now need to add result type rank check for conv2DBackpropInputOp. Right?

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1246,6 +1218,121 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+template <typename OpT,
+          typename std::enable_if<llvm::is_one_of<
+              OpT, Conv2DOpAdaptor, Conv3DOpAdaptor>::value>::type* = nullptr>
+static LogicalResult verifyConvReturnTypes(
```
nit: this is not verifying the retrun types but inferring them so we should rename this function.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1246,6 +1218,121 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+  return_shape[GetTensorBatchDimIndex(num_dims, format)] =
+      input_ty.getShape()[GetTensorBatchDimIndex(num_dims, format)];
+  return_shape[GetTensorFeatureDimIndex(num_dims, format)] =
+      std::is_same<OpT, Conv2DOpAdaptor>() ? filter_ty.getShape()[3]
```
optional nit: Instead of using magic constants, we should use GetFilterTensorInnerInputChannelsDimIndex here.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1129,80 +1131,50 @@ LogicalResult ConstOp::inferReturnTypes(
   int num_spatial_dims = std::is_same<OpT, Conv2DOp>() ? 2 : 3;
   int num_dims = 2 + num_spatial_dims;
 
-  if (!IsOfRankOrUnranked(op.input(), num_dims) ||
```
I think, we still need to keep these checks on operand ranks. Move it to a shared function.

pr4tgpt replied:
Yes This is present in the ``inferConvReturnTypes()`` function

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1129,80 +1131,50 @@ LogicalResult ConstOp::inferReturnTypes(
-  // EXPLICIT padding mode and the associated attribute is limited to Conv2D.
-  // So, fetch attribute by string instead of the op.explicit_paddings()
-  // attribute getter.
-  if (op.padding() == "EXPLICIT") {
```
We should move this verification of padding attribute in VerifyConvOpAttributes for sharing.

pr4tgpt replied:
Yes This is present in the ``inferConvReturnTypes()`` function

smit-hinsu reviewed (approved):
Thanks for all the hard work here. We are almost there!

Let me know if any of the comments are not clear and you need further pointer/clarification.

gbaned commented:
@pr4tgpt  Can you please check @smit-hinsu's comments and keep us posted ? Thanks!

pr4tgpt commented:
> @pr4tgpt Can you please check @smit-hinsu's comments and keep us posted ? Thanks!

I have checked and resolved the comments. Thanks

pr4tgpt commented:
Hi @smit-hinsu , I have made changes as requested by you. Please review the changes and feel free to post suggestions/improvements. Thanks!

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1446,6 +1418,140 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+  explicit_padding = explicit_pad.getValue();
+  StringRef filterFormat;
+  StringAttr filterFormatAttr =
+      attributes.get("filter_format").dyn_cast_or_null<::mlir::StringAttr>();
```
These ops doesn't have filter_format attribute so this will always fail to fetch the attribute.

pr4tgpt replied:
Yes, That's true, but for completeness, I did this.

smit-hinsu replied:
It is not good idea to have dead code. Let's keep it as it is.

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -2493,4 +2630,4 @@ OpFoldResult MulOp::fold(ArrayRef<Attribute> operands) {
 
 #define GET_OP_CLASSES
-#include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc.inc"
+#include "tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc.inc"
```
nit: add the newline at the end to avoid diff here.

smit-hinsu reviewed (approved):
Looks perfect after couple of minor modifications.

smit-hinsu reviewed (changes requested):
While updating the PR, consider adding new commits instead of amending it so that it is easier to see the changes.

pr4tgpt committed:
Removed dead code and duplicate verification functions.
In this commit, some of the dead code is removed. Also a duplicate function for verifying convolution attributes is also removed.

Signed-off-By: Prateek Gupta<prateek@polymagelabs.com>

pr4tgpt commented:
> While updating the PR, consider adding new commits instead of amending it so that it is easier to see the changes.

Hi @smit-hinsu I have made changes as suggested by you in a new commit. Please review it. Thanks!

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1446,6 +1406,134 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+      input_ty.getShape()[GetTensorBatchDimIndex(num_dims, format)];
+  return_shape[GetTensorFeatureDimIndex(num_dims, format)] =
+      filter_ty.getShape()[GetFilterTensorInnerInputChannelsDimIndex(
+          num_dims, filterFormat)];
```
nit: we should directly use FORMAT_HWIO here and drop the argument.

pr4tgpt replied:
Hi @smit-hinsu changed this. Thanks!

pr4tgpt committed:
Add correction in some of the test cases.
Some of the test cases were using wrong output sizes for tf.conv2D. This commit corrects those test cases.

Signed-off-by: Prateek Gupta<prateek@polymagelabs.com>

pr4tgpt commented:
Hi, @smit-hinsu Some of the newer test cases were using wrong output sizes for ``tf.conv2d``. I have corrected them. Please review that. Thanks!

pr4tgpt commented:
Hi, @smit-hinsu some more failing test cases corrected. Please review them also. Thanks!

pr4tgpt commented:
Hi, the code has been approved. Please merge this after running checks thanks.

pr4tgpt commented:
Hi @smit-hinsu, Can you look into why the internal build is failing here?  I cannot access the build report. Thanks!

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1467,6 +1427,129 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+  tensorflow::TensorFormat format;
+  FormatFromString(data_format.str(), &format);
+  tensorflow::Padding padding;
+  GetPaddingFromString(paddings.str(), &padding);
```
we need to do error handling here as the function has warn_unused_result.

pr4tgpt replied:
Hi @smit-hinsu. Added the checks. Please review. Thanks!

pr4tgpt committed:
Add minor checks for data_format and padding value

smit-hinsu commented on tensorflow/compiler/mlir/tensorflow/ir/tf_ops_a_m.cc:
```diff
@@ -1467,6 +1427,131 @@ LogicalResult Conv2DOp::UpdateDataFormat(StringRef data_format) {
+
+  tensorflow::TensorFormat format;
+  auto data_format_is_valid = FormatFromString(data_format.str(), &format);
+  DCHECK(data_format_is_valid) << data_format.str();
```
emit an optional error instead of DCHECK to handle error cases.

pr4tgpt replied:
@smit-hinsu, Done. Please review. Thanks!

pr4tgpt commented:
Hi @smit-hinsu, can you look into what error has occurred? Thanks!

smit-hinsu commented:
Could you run all the tests in these directories and fix the failing ones?

tensorflow/compiler/mlir
tensorflow/compiler/tests
tensorflow/lite/python
tensorflow/lite/testing/model_coverage

gbaned commented:
@pr4tgpt Can you please check @smit-hinsu's comments and keep us posted ? Thanks!

pr4tgpt commented:
> Could you run all the tests in these directories and fix the failing ones?
> 
> tensorflow/compiler/mlir
> tensorflow/compiler/tests
> tensorflow/lite/python
> tensorflow/lite/testing/model_coverage

Yes I am looking into this. Thanks!

pr4tgpt commented:
> Could you run all the tests in these directories and fix the failing ones?
> 
> tensorflow/compiler/mlir
> tensorflow/compiler/tests
> tensorflow/lite/python
> tensorflow/lite/testing/model_coverage

Hi @smit-hinsu I checked the tests you asked. 

The tests in ``tensorflow/compiler/mlir`` specific to ``tf_to_gpu`` and ``tf_to_kernel`` pass fail. These tests have no dependency to my patch.
The ``python`` tests in the ``tensorflow/compiler/tests`` are failing due to reasons I can't determine(mostly ``ModuleNameError``). I find no dependency to my patch here also.
The ``python`` tests in the ``tensorflow/lite/python`` are failing due to reasons I can't determine(mostly ``ImportError``). I find no dependency to my patch here also.
The tests in ``tensorflow/lite/testing/model_coverage`` pass successfully.

Could you see the failing tests once and tell me what could be the possible reason for them to fail? Thanks!

pr4tgpt commented:
> > Could you run all the tests in these directories and fix the failing ones?
> > tensorflow/compiler/mlir
> > tensorflow/compiler/tests
> > tensorflow/lite/python
> > tensorflow/lite/testing/model_coverage
> 
> Hi @smit-hinsu I checked the tests you asked.
> 
> The tests in `tensorflow/compiler/mlir` specific to `tf_to_gpu` and `tf_to_kernel` pass fail. These tests have no dependency to my patch.
> The `python` tests in the `tensorflow/compiler/tests` are failing due to reasons I can't determine(mostly `ModuleNameError`). I find no dependency to my patch here also.
> The `python` tests in the `tensorflow/lite/python` are failing due to reasons I can't determine(mostly `ImportError`). I find no dependency to my patch here also.
> The tests in `tensorflow/lite/testing/model_coverage` pass successfully.
> 
> Could you see the failing tests once and tell me what could be the possible reason for them to fail? Thanks!

Hi @smit-hinsu could you help with this? Thanks!

smit-hinsu commented:
All these tests depend on your change. tensorflow/compiler/mlir/tensorflow/tests uses tf-opt tool which links everything andd tensorflow/lite/python links the TF dialect so there are dependencies.
 
There are many tests failing. Following are some of these:

tensorflow/compiler/mlir/tensorflow/tests:tf-ops.mlir.test
tensorflow/compiler/mlir/lite/tests:prepare-tf.mlir.test
tensorflow/compiler/tests:conv3d_mlir_bridge_test_cpu
tensorflow/lite/python:lite_test
tensorflow/lite/testing/model_coverage:model_coverage_test_user_vgg_tflite_builtins_select_tf_ops

Please take a look. You can use 'tensorflow/compiler/mlir/tensorflow/tests:all' to test all tests in the directory.

bondhugula commented:
These failures are all clearly related to this PR. An example from `tensorflow/compiler/mlir/tensorflow/tests`:

```
2020-12-10 21:49:06.560035: F ./tensorflow/core/util/tensor_format.h:266] Check failed: format == FORMAT_OIHW_VECT_I (0 vs. 3)
TensorFlow crashed, please file a bug on https://github.com/tensorflow/tensorflow/issues with the trace below.
Stack dump:
0.      Program arguments: /home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tf-opt /home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir -tf-optimize 
1.      Program arguments: /home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tf-opt /home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir -tf-optimize 
 #0 0x00000000122c096f llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tf-opt+0x122c096f)
 #1 0x00000000122be5ed llvm::sys::RunSignalHandlers() (/home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tf-opt+0x122be5ed)
 #2 0x00000000122bee6d SignalHandler(int) (/home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tf-opt+0x122bee6d)
 #3 0x00007fe1f5214dd0 __restore_rt (/lib64/libpthread.so.0+0x12dd0)
 #4 0x00007fe1f4c5f70f raise (/lib64/libc.so.6+0x3770f)
 #5 0x00007fe1f4c49b25 abort (/lib64/libc.so.6+0x21b25)
 #6 0x00000000111d2b10 virtual thunk to tensorflow::internal::LogMessageFatal::~LogMessageFatal() (/home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-o
 #7 0x000000000fe767ec mlir::LogicalResult mlir::TF::inferConvReturnTypes<mlir::TF::Conv2DOpAdaptor, (void*)0>(mlir::TF::Conv2DOpAdaptor, llvm::SmallVectorImpl<mlir::Type>&, llvm::Optional<mlir::Location>, llvm::ArrayRef<mlir::Attribute>) (.cold.3446) (/home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tf-opt+0xfe767ec)
 #8 0x000000000fe77224 mlir::TF::Conv2DOp::inferReturnTypes(mlir::MLIRContext*, llvm::Optional<mlir::Location>, mlir::ValueRange, mlir::DictionaryAttr, mlir::RegionRange, llvm::SmallVectorImpl<mlir::Type>&) (/home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tf-opt+0xfe77224)
 #9 0x0000000011eeb238 mlir::detail::verifyInferredResultTypes(mlir::Operation*) (/home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tf-opt+0x11eeb238)
#10 0x000000000fc884d9 mlir::Op<mlir::TF::Conv2DOp, mlir::OpTrait::ZeroRegion, mlir::OpTrait::OneResult, mlir::OpTrait::ZeroSuccessor, mlir::OpTrait::NOperands<2u>::Impl, mlir::InferTypeOpInterface::Trait, mlir::MemoryEffectOpInterface::Trait, mlir::TF::LayoutSensitiveInterface::Trait, mlir::DerivedAttributeOpInterface::Trait>::verifyInvariants(mlir::Operation*) (/home/uday/.cache/bazel/_bazel_uday/97da8b8d2f64c0a54677e92074ce1d42/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/mlir/tensorflow/tests/tf_optimize.mlir.test.runfiles/org_tensorflow/tensorflow/compiler/mlir/tf-opt+0xfc884d9)


```

pr4tgpt committed:
[MLIR] Add minor fix for filter dimensions.
This commit addresses a bug where incorrect filter dimensions were being read and used. Corrected the bug and the relevant test cases were also checked.

Signed-off-by: Prateek Gupta <prateek@polymagelabs.com>

pr4tgpt commented:
> All these tests depend on your change. tensorflow/compiler/mlir/tensorflow/tests uses tf-opt tool which links everything andd tensorflow/lite/python links the TF dialect so there are dependencies.
> 
> There are many tests failing. Following are some of these:
> 
> tensorflow/compiler/mlir/tensorflow/tests:tf-ops.mlir.test
> tensorflow/compiler/mlir/lite/tests:prepare-tf.mlir.test
> tensorflow/compiler/tests:conv3d_mlir_bridge_test_cpu
> tensorflow/lite/python:lite_test
> tensorflow/lite/testing/model_coverage:model_coverage_test_user_vgg_tflite_builtins_select_tf_ops
> 
> Please take a look. You can use 'tensorflow/compiler/mlir/tensorflow/tests:all' to test all tests in the directory.

Hi @smit-hinsu, @bondhugula  I have fixed the code. The filter dimensions were fetched incorrectly. The ``tensorflow/compiler/mlir/tensorflow/tests`` are now passing. 
Looking at other failing tests also. Thanks! Please review.

pr4tgpt commented:
> > All these tests depend on your change. tensorflow/compiler/mlir/tensorflow/tests uses tf-opt tool which links everything andd tensorflow/lite/python links the TF dialect so there are dependencies.
> > There are many tests failing. Following are some of these:
> > tensorflow/compiler/mlir/tensorflow/tests:tf-ops.mlir.test
> > tensorflow/compiler/mlir/lite/tests:prepare-tf.mlir.test
> > tensorflow/compiler/tests:conv3d_mlir_bridge_test_cpu
> > tensorflow/lite/python:lite_test
> > tensorflow/lite/testing/model_coverage:model_coverage_test_user_vgg_tflite_builtins_select_tf_ops
> > Please take a look. You can use 'tensorflow/compiler/mlir/tensorflow/tests:all' to test all tests in the directory.
> 
> Hi @smit-hinsu, @bondhugula I have fixed the code. The filter dimensions were fetched incorrectly. The `tensorflow/compiler/mlir/tensorflow/tests` are now passing.
> Looking at other failing tests also. Thanks! Please review.

Tests in ``mlir/lite/tests`` are also passing.

smit-hinsu commented:
It finally got submitted! Thanks Prateek for the contribution and working through all the issues.

pr4tgpt commented:
> It finally got submitted! Thanks Prateek for the contribution and working through all the issues.

Thank you Smit for your valuable help and suggestions.
